code/game/atom/atom_color.dm

```diff
@@ -43,29 +51,62 @@
  */
 /atom/proc/is_atom_colour(looking_for_color, min_priority_index = 1, max_priority_index = COLOUR_PRIORITY_AMOUNT)
 	// make sure uppertext hex strings don't mess with LOWER_TEXT hex strings
-	looking_for_color = LOWER_TEXT(looking_for_color)
+	if (!islist(looking_for_color))
+		looking_for_color = LOWER_TEXT(looking_for_color)
 
 	if(!LAZYLEN(atom_colours))
 		// no atom colors list has been set up, just check the color var
-		return LOWER_TEXT(color) == looking_for_color
+		if (!islist(color))
+			return LOWER_TEXT(color) == looking_for_color
+		if (!islist(looking_for_color))
+			return FALSE
+		return compare_list(color, looking_for_color)
 
 	for(var/i in min_priority_index to max_priority_index)
-		if(LOWER_TEXT(atom_colours[i]) == looking_for_color)
+		if (atom_colours[i][ATOM_COLOR_TYPE_INDEX] == ATOM_COLOR_TYPE_NORMAL)
+			if(LOWER_TEXT(atom_colours[i][ATOM_COLOR_VALUE_INDEX]) == looking_for_color)
+				return TRUE
+			continue
+
+		if (!islist(looking_for_color))
+			continue
+
+		if (compare_list(looking_for_color, atom_colours[i][ATOM_COLOR_VALUE_INDEX]["color"]))
 			return TRUE
 
 	return FALSE
 
 ///Resets the atom's color to null, and then sets it to the highest priority colour available
 /atom/proc/update_atom_colour()
 	color = null
-	if(!atom_colours)
+	remove_filter(FILTER_ATOM_PRIORITY_COLOR)
+
+	if (!atom_colours)
 		return
-	for(var/checked_color in atom_colours)
-		if(islist(checked_color))
-			var/list/color_list = checked_color
-			if(color_list.len)
-				color = color_list
-				return
-		else if(checked_color)
-			color = checked_color
+
+	for (var/list/checked_color in atom_colours)
+		if (checked_color[ATOM_COLOR_TYPE_INDEX] == ATOM_COLOR_TYPE_FILTER)
+			add_filter(FILTER_ATOM_PRIORITY_COLOR, 1, checked_color[ATOM_COLOR_VALUE_INDEX])
+			return
+
+		if (length(checked_color[ATOM_COLOR_VALUE_INDEX]))
+			color = checked_color[ATOM_COLOR_VALUE_INDEX]
+			return
+
+/// Same as update_atom_color, but simplifies overlay coloring
+/atom/proc/color_atom_overlay(mutable_appearance/overlay)
+	overlay.color = null
+	overlay.remove_filter(FILTER_ATOM_PRIORITY_COLOR)
+
+	if (!atom_colours)
+		overlay.color = color
+		return
+
+	for (var/list/checked_color in atom_colours)
+		if (checked_color[ATOM_COLOR_TYPE_INDEX] == ATOM_COLOR_TYPE_FILTER)
+			overlay.add_filter(FILTER_ATOM_PRIORITY_COLOR, 1, checked_color[ATOM_COLOR_VALUE_INDEX])
+			return
+
+		if (length(checked_color[ATOM_COLOR_VALUE_INDEX]))
+			overlay.color = checked_color[ATOM_COLOR_VALUE_INDEX]
 			return
```

Comment:
the fact that we need this is concerning, was it an issue with the old system as well?

Reply:
Overlays inherit color but not filters, thus in some cases you'd want to color them (locker doors)

Reply:
It initially was made to easily copy overlays onto worn items, but seeing how its an issue on closets I'm pretty sure this applies to literally everything (don't know if they have KEEP_TOGETHER or not, but none of their parents explicitly define appearance_flags)

Reply:
that's a real problem then, we're introducing a footgun here that only shows up rarely. many things are built off composite overlays after all

Reply:
...A very ugly, but viable solution would be adding the filter to all overlays in update_overlays, but I am not sure how bad that's gonna be performance-wise